x-pack/plugins/rule_registry/server/event_log/log/event_log_resolver.ts

```diff
@@ -0,0 +1,136 @@
+import { FieldMap } from '../event_schema';
+import { EventLogBootstrapper } from './event_log_bootstrapper';
+import { EventLogProvider } from './event_log_provider';
+import { IEventLogRegistry } from './internal_api';
+import {
+  EventLogServiceConfig,
+  EventLogServiceDependencies,
+  IEventLogDefinition,
+  IEventLogProvider,
+  IEventLogResolver,
+} from './public_api';
+import { mappingFromFieldMap } from './utils/mapping_from_field_map';
+
+export class EventLogResolver implements IEventLogResolver {
+  private readonly indexBootstrapper: IndexBootstrapper;
+
+  constructor(
+    private readonly config: EventLogServiceConfig,
+    private readonly deps: EventLogServiceDependencies,
+    private readonly registry: IEventLogRegistry
+  ) {
+    this.indexBootstrapper = this.createIndexBootstrapper();
+  }
+
+  resolve<TMap extends FieldMap>(
+    definition: IEventLogDefinition<TMap>,
+    kibanaSpaceId: string
+  ): IEventLogProvider<TMap> {
+    const existingProvider = this.registry.get(definition, kibanaSpaceId);
+    if (existingProvider) {
+      return existingProvider;
+    }
+
+    const indexSpec = this.createIndexSpec(definition, kibanaSpaceId);
+    const indexReader = this.createIndexReader(indexSpec);
+    const indexWriter = this.createIndexWriter(indexSpec);
+    const logBootstrapper = this.createEventLogBootstrapper(indexSpec);
+    const logProvider = new EventLogProvider<TMap>({
+      eventSchema: definition.eventSchema,
+      indexSpec,
+      indexReader,
+      indexWriter,
+      logBootstrapper,
+      logger: this.deps.logger,
+    });
+
+    const closeLog = async (): Promise<void> => {
+      await indexWriter.close();
+    };
+
+    this.registry.add(definition, kibanaSpaceId, logProvider, closeLog);
+
+    return logProvider;
+  }
+
+  private createIndexSpec<TMap extends FieldMap>(
+    definition: IEventLogDefinition<TMap>,
+    kibanaSpaceId: string
+  ): IndexSpecification {
+    const { indexPrefix } = this.config;
+    const { eventLogName, eventSchema, ilmPolicy } = definition;
+
+    const indexNames = IndexNames.create({
+      indexPrefix,
+      logName: eventLogName,
+      kibanaSpaceId,
+    });
+
+    const indexMappings = mappingFromFieldMap(eventSchema.objectFields);
+
+    return { indexNames, indexMappings, ilmPolicy };
+  }
+
+  private createIndexBootstrapper(): IndexBootstrapper {
+    const { clusterClient, logger } = this.deps;
+
+    return new IndexBootstrapper({
+      gateway: new IndexManagementGateway({
+        elasticsearch: clusterClient.then((c) => c.asInternalUser),
+        logger,
+      }),
+      logger,
+    });
+  }
+
+  private createIndexReader(indexSpec: IndexSpecification): IndexReader {
+    const { clusterClient, logger } = this.deps;
+    const { indexNames } = indexSpec;
+
+    return new IndexReader({
+      indexName: indexNames.indexAliasPattern,
+      elasticsearch: clusterClient.then((c) => c.asInternalUser), // TODO: internal or current?
```

Comment:
BTW, I'm getting the following error during the index bootstrapping:
```
Error: Event log bootstrapping failed, logName="security.alerts"
    at EventLogProvider.getLog (/Users/myusers/elastic/kibana-master/x-pack/plugins/rule_registry/server/event_log/log/event_log_provider.ts:68:15)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at testImplementation (/Users/myuser/elastic/kibana-master/x-pack/plugins/rule_registry/server/event_log/test_implementation.ts:72:3)
    at /Users/myuser/elastic/kibana-master/x-pack/plugins/rule_registry/server/event_log/test_implementation.ts:42:7
    ```
    I haven't had a chance to debug it yet, but it looks like it could be related to this line.

Reply:
I think most likely this is caused by the fact that you have `kibana_system` as your internal user specified in the config.

https://github.com/elastic/elasticsearch/pull/72181 has fixed that. I'll rebase the PR and hopefully after that the implementation will work with `kibana_system` user.